de.dlr.sc.virsat.model.test/src/de/dlr/sc/virsat/model/dvlm/tree/TreeTraverserTest.java

```diff
@@ -0,0 +1,169 @@
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+public class TreeTraverserTest {
+
+	StructuralElement se;
+	
+	StructuralElementInstance seiRoot;
+	StructuralElementInstance seiChild1;
+	StructuralElementInstance seiChild1Child;
+	StructuralElementInstance seiChild1ChildChild;
+	StructuralElementInstance seiChild2;
+	StructuralElementInstance seiChild2Child;
+	StructuralElementInstance seiChild2ChildChild;
+	
+	@Before
+	public void setUp() throws Exception {
+		StructuralElement se = StructuralFactory.eINSTANCE.createStructuralElement();
+		
+		seiRoot = StructuralFactory.eINSTANCE.createStructuralElementInstance();
+		seiChild1 = StructuralFactory.eINSTANCE.createStructuralElementInstance();
+		seiChild1Child = StructuralFactory.eINSTANCE.createStructuralElementInstance();
+		seiChild1ChildChild = StructuralFactory.eINSTANCE.createStructuralElementInstance();
+		seiChild2 = StructuralFactory.eINSTANCE.createStructuralElementInstance();
+		seiChild2Child = StructuralFactory.eINSTANCE.createStructuralElementInstance();
+		seiChild2ChildChild = StructuralFactory.eINSTANCE.createStructuralElementInstance();
+		
+		se.setIsApplicableForAll(true);
+		
+		seiRoot.setType(se);
+		seiChild1.setType(se);
+		seiChild1Child.setType(se);
+		seiChild1ChildChild.setType(se);
+		seiChild2.setType(se);
+		seiChild2Child.setType(se);
+		seiChild2ChildChild.setType(se);
+		
+		seiRoot.getChildren().add(seiChild1);
+		seiRoot.getChildren().add(seiChild2);
+		seiChild1.getChildren().add(seiChild1Child);
+		seiChild2.getChildren().add(seiChild2Child);
+		seiChild1Child.getChildren().add(seiChild1ChildChild);
+		seiChild2Child.getChildren().add(seiChild2ChildChild);
+		
+		// Now set names for the 2 level and for the 3 level,
+		// Try to find them and check if the traverser reacts as expected
+		seiRoot.setName("Root");
+		seiChild1Child.setName("ChildOf1");
+		seiChild2Child.setName("ChildOf2");
+		seiChild1ChildChild.setName("DoubleChildOf1");
+		seiChild2ChildChild.setName("DoubleChildOf2");
+	}
+
+	/**
+	 * Class to capture the results from the tree traversing
+	 */
+	class Result {
+		
+		@Override
+		public int hashCode() {
+			final int prime = 31;
+			int result = 1;
+			result = prime * result + Objects.hash(matchingLevel, nestingLevel, sei);
+			return result;
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			if (this == obj) {
+				return true;
+			} else if (obj == null) {
+				return false;
+			} else if (getClass() != obj.getClass()) {
+				return false;
+			} else {
+				Result other = (Result) obj;
+				return matchingLevel == other.matchingLevel && nestingLevel == other.nestingLevel && Objects.equals(sei, other.sei);
+			}
+		}
+		
+		Result(StructuralElementInstance sei, int processedLevel, int matchedLevel) {
+			this.sei = sei;
+			this.nestingLevel = processedLevel;
+			this.matchingLevel = matchedLevel;
+		}

```

Comment:
Can we move these to the top of the `Result` class so the reader directly knows whats encapsualted by it?

Reply:
yes done.